### PR TITLE
Core: Validate concurrently added delete files in RowDelta

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR validates concurrently added delete files in `BaseRowDelta`.

### CODE REVIEW

File: api/src/main/java/org/apache/iceberg/RowDelta.java [hunk 1]
```diff
@@ -109,6 +109,54 @@
...
+
+  /**
+   * Sets a conflict detection filter used to validate concurrently added data and delete files.
+   * <p>
+   * If not called, a true literal will be used as the conflict detection filter.
+   *
+   * @param conflictDetectionFilter an expression on rows in the table
+   * @return this for method chaining
+   */
+  RowDelta conflictDetectionFilter(Expression conflictDetectionFilter);
```
Comment (aokolnychyi) [inline api/src/main/java/org/apache/iceberg/RowDelta.java:129]: This is actually different compared to the implementation in PR #3069. I think this one is slightly better as we set the conflict detection filter only once and then enable data/delete file validation.

**Comment by aokolnychyi** [comment]
cc @openinx @rdblue @RussellSpitzer @jackye1995 @yyanyy @flyrain @kbendick @karuppayya @raptond @szehon-ho

File: core/src/main/java/org/apache/iceberg/DeleteFileIndex.java [hunk 2]
```diff
@@ -86,6 +86,20 @@ public boolean isEmpty() {
     return (globalDeletes == null || globalDeletes.length == 0) && sortedDeletesByPartition.isEmpty();
   }
 
+  public List<DeleteFile> referencedDeleteFiles() {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/DeleteFileIndex.java:89]: Would `Iterable<DeleteFile>` work instead? That would make this lazy.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java [hunk 3]
```diff
@@ -320,6 +333,56 @@ protected void validateNoNewDeletesForDataFiles(TableMetadata base, Long startin
...
+   *
+   * @param base table metadata to validate
+   * @param startingSnapshotId id of the snapshot current at the start of the operation
+   * @param dataFilter an expression used to find new conflicting delete files
+   * @param caseSensitive whether expression evaluation should be case-sensitive
+   */
+  protected void validateNoNewDeleteFiles(TableMetadata base, Long startingSnapshotId,
+                                          Expression dataFilter, boolean caseSensitive) {
+    // if there is no current table state, no files have been added
+    if (base.currentSnapshot() == null) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java:347]: Can you add the check for `base.formatVersion() < 2`?

File: core/src/main/java/org/apache/iceberg/DeleteFileIndex.java (hunk 2 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/DeleteFileIndex.java]: Updated.

File: core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java]: Added to both methods. Could you check, @rdblue?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good. Thanks for adding that check!

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @rdblue!

**Pull request merged**